rsocket/RSocketServiceHandler.h

```diff
@@ -0,0 +1,97 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+#include "rsocket/RSocketException.h"
+#include "rsocket/RSocketNetworkStats.h"
+#include "rsocket/RSocketParameters.h"
+#include "rsocket/RSocketResponder.h"
+#include "rsocket/RSocketServerState.h"
+#include "rsocket/RSocketStats.h"
+#include "rsocket/internal/Common.h"
+
+namespace rsocket {
+
+// This is a convenience function which is used to create a simplified
+// RSocketServiceHandler.
+using OnNewSetupFn =
+    folly::Function<std::shared_ptr<RSocketResponder>(const SetupParameters&)>;
+
+// This struct holds all the necessary information needed by the RSocketServer
+// to initiate a connection with a client.
+struct RSocketConnectionParams {
+  RSocketConnectionParams(
+      std::shared_ptr<RSocketResponder> _responder,
+      std::shared_ptr<RSocketStats> _stats = RSocketStats::noop(),
+      std::shared_ptr<RSocketNetworkStats> _networkStats = nullptr)
+      : responder(std::move(_responder)),
+        stats(std::move(_stats)),
+        networkStats(std::move(_networkStats)) {}
+  std::shared_ptr<RSocketResponder> responder;
+  std::shared_ptr<RSocketStats> stats;
+  std::shared_ptr<RSocketNetworkStats> networkStats;
+};
+
+// This class has to be implemented by the application.  The methods can be
+// called from different threads and it is the application's responsibility to
+// ensure thread-safety.
+class RSocketServiceHandler {
+ public:
+  virtual ~RSocketServiceHandler() = default;
+
+  // This method gets called for each client that connects to the server.  The
+  // application has to implement this method.  If the application does not
+  // want to accept this connection, it should return a RSocketException.  If
+  // the application returns a RSocketException, then an ERROR frame with code
+  // REJECTED_SETUP is sent to the client.  The exception message is sent as
+  // payload.
+  virtual folly::Expected<RSocketConnectionParams, RSocketException> onNewSetup(
+      const SetupParameters&) = 0;
```

Comment:
The ownership of SetupParams still resides with the protocol-implementation.  I did not want to move it away to the application.  The RSocketStateMachine::connectServer() for example, takes the SetupParams as parameter (though it does not use the initialPayload).  My reasoning around this has been, though initialPayload is not currently being used by the current protocol-implementation, moving in the direction where we lose this information could leave us handicapped later when we want to use it somewhere.